### ISSUE TITLE
Core: ability to launch a group of requests in parallel outside the resources scope

### ISSUE THREAD

**Issue by notdryft** [gatling/gatling#3783]
Single Page application often perform multiple requests at the same time while viewing pages.

Right now, the only way to simulate to behavior is to use the `resources` DSL element. However, if you do so you'd end up silencing your precious API requests with `silentResources`...

I'd see multiple ways to provide this ability :

- A copy of the `resources` DSL that isn't silenced
- A full-fledged DSL element that enables users to have parallel workflows inside the simulation like we do with WebSockets, Polling or SSE

See #2336 for a previous query on this

**Comment by phiSgr** [comment]
How about this?

```scala
def async(attributeName: String, actionBuilder: ActionBuilder): ActionBuilder = new ActionBuilder {
  override def build(ctx: ScenarioContext, next: Action): Action = new Action with NameGen {
    override def name: String = genName("async")

    override def execute(session: Session): Unit = {
      val branched = Promise[Session]()
      actionBuilder.build(ctx, new Action {
        override def name: String = genName("asyncFiller")

        override def execute(session: Session): Unit = branched.success(session)
      }).execute(session)
      next ! session.set(attributeName, branched.future)
    }
  }
}

def await(attributeName: String)(combiner: (Session, Session) => Session): ActionBuilder = new ActionBuilder {
  override def build(ctx: ScenarioContext, _next: Action): Action = new ChainableAction with NameGen {
    override val next: Action = _next

    override def name: String = genName("await")

    override def execute(session: Session): Unit = {
      implicit val ec = ctx.coreComponents.actorSystem.dispatcher
      val f = session.attributes(attributeName).asInstanceOf[Future[Any]]
      f.onComplete {
        case Success(branched: Session) =>
          val mainFlow = session.remove(attributeName)
          val newSession = combiner(branched, if (branched.isFailed) mainFlow.markAsFailed else mainFlow)
          next ! newSession
        case t =>
          logger.error(s"'$name' crashed because of $attributeName does not contain a Future[Session], " +
            "forwarding to the next one")
          next ! session.markAsFailed
      }
    }

  }
}
```

**Comment by pschiffe** [comment]
This feature could probably solve this problem as well:
https://groups.google.com/forum/#!topic/gatling/C2cKGviiuKI

**Comment by pschiffe** [comment]
When going through the mailing list, I see more topics regarding this feature request:

https://groups.google.com/d/msg/gatling/655oWQFf72o/mXfOgUNiFgAJ
https://groups.google.com/d/msg/gatling/pfOECwG-mwA/X3WrRFK2AQAJ
https://groups.google.com/d/msg/gatling/hAWKTn6ONBc/f_1wCa8pjsYJ
https://groups.google.com/d/msg/gatling/N4PajpAuBMg/oQHcWFgUBQAJ
https://groups.google.com/d/msg/gatling/Q_hg8PYbc5U/nGbX59ejBwAJ
https://groups.google.com/d/msg/gatling/EmcPTxV0BlA/4luDCbnJWm0J
https://groups.google.com/d/msg/gatling/Rn1Ev1VW3X0/IhihWEpmCQAJ
https://groups.google.com/d/msg/gatling/ujlGuBeAObI/npzqG06oFAAJ
https://groups.google.com/d/msg/gatling/eAgsDbvZna8/3qRPvquSDAAJ
https://groups.google.com/d/msg/gatling/vr6GknCI7r4/QtIguhE81tUJ
https://groups.google.com/d/msg/gatling/1kGyMC_0DyQ/H202S_GxBwAJ

**Comment by slandelle** [comment]
Open source: if you really need something, the best way to make it happen is to contribute it.

**Comment by pschiffe** [comment]
Yes, I'm familiar. I'm contributing by gathering user feedback and reporting bugs, unfortunately I cannot contribute Scala code, I would if I could.

**Comment by zawawimanja** [comment]
any update on this ?

**Comment by slandelle** [comment]
Ticket is still open without a milestone set, so answer is likely no.

**Comment by robstoll** [comment]
@slandelle I am also interested in this feature and might contribute it. However, I would like to know first if there are reasons that this wasn't implemented already (I did not contribute to gatling yet, so maybe there are some restrictions I should be aware of etc.). 
For me the `resources` hack doesn't work as I need cascading behaviour, i.e. run some requests in parallel and for some of them, run again requests in parallel once their are done etc. Also, I use the `.silentResources()` option in the HttpProtocolBuilder but would like to see those requests in the report.

**Comment by slandelle** [comment]
> I am also interested in this feature and might contribute it

Great! Please make sure to share your design first (DSL, implementation), before coding :)

> However, I would like to know first if there are reasons that this wasn't implemented already

Time, effort, priority.

> the resources hack doesn't work as I need cascading behaviour,

Well, it's honestly very complicated. First, you need a DSL to describe this cascading. Then, you need a FSM to implement it, and take the max concurrent number of HTTP/1.1 connections/HTTP/2 streams into account.

> Also, I use the .silentResources() option in the HttpProtocolBuilder but would like to see those requests in the report.

I don't follow. The whole point of `silentResources` is to make them not appear in the reports.

> ps: I only plan to contribute to scala/java but not typescript, would that be a problem?

No, we can take care of the JS bridge.

Good luck!

**Comment by robstoll** [comment]
> I don't follow. The whole point of silentResources is to make them not appear in the reports.

Exactly, and since I want my parallel requests to appear in the report, I cannot use `resources`. 

>  take the max concurrent number of HTTP/1.1 connections/HTTP/2 streams into account.

Thanks, that's something I did not thought about and it even depends on the browser in use and if the requests are for the same domain or not. I would keep it simple for a start and not impose any limits (and let the user deal with it if one wants). We can then go from there to some default limit overall which can be overridden (via config) and then to per domain. 

Also, I see that this can get more complex if we want to support the execution of all actions in parallel. I suggest we start off with only supporting HttpRequestAction.

WDYT?

**Comment by slandelle** [comment]
I'd say that before jumping into the implementation, the first step is to define the DSL. How would you define parallel requests?

**Comment by robstoll** [comment]
DSL wise I would imagine something like:
```
scenario..
  .exec(a) // as is, i.e. sequential
  .exec(b) // b executes once a is done
  .execParallel( // same signature as exec, for things being a ChainBuilder you can again use execParallel
     // for instance, 3 exec which are executed in parallel
     exec(c)
     exec(d)
     exec(e).execParallel( // starts parallel execution of f and g once e is done
       exec(f)
       exec(g)
     )
  )
```

**Comment by slandelle** [comment]
That won't work. What you're suggesting in not concurrent requests, it's concurrent pieces of scenario. This is not currently possible with Gatling 3. IMO, we should only focus on HTTP requests.

How do you model "start f and g once c, d and e are done"?

**Comment by slandelle** [comment]
> You misread I guess

I didn't. I described another kind of flow other people will probably want. We have to design a DSL that can handle this case too.

`http(e).andThen` doesn't work. `http` won't have an `andThen` method as it only makes sense inside an `httpParallel` block.

The hardest part is the DSL design ;)

**Comment by robstoll** [comment]
> I didn't.

alright, then "start f and g once c, d and e are done"?
```
exec(httpParallel(c, e, e)).exec(httpParallel(f, g))
```
> http(e).andThen doesn't work. http won't have an andThen method as it only makes sense inside an httpParallel block.

Although I agree that the intention would be that `andThen` is used inside `httpParallel`, I don't see how this should not work. What would be the semantic difference between it and `exec(http(a)).exec(http(b))` => b will be executed after a in both cases
